examples/Auction/StateMachine.hs

```diff
@@ -0,0 +1,339 @@
+      gen s =
+        case filter (enoughForFees . fst . snd) $ Map.toList (users s) of
+          [] -> Nothing
+          us -> Just $ Start . second snd <$> Gen.element us
+
+      execute :: Start Concrete -> PropertyT RunIO TxOutRef
+      execute (Start (_, Var (Concrete pkh))) = liftRun $ PSM.start pkh
+   in Command
+        gen
+        execute
+        [ Update $ \s (Start (user, _)) o ->
+            s
+              { currentBid = Just (user, user, 0, o)
+              }
+        , Require $ \input (Start (u, _)) ->
+            case Map.lookup u (users input) of
+              Just (bal, _) -> enoughForFees bal
+              Nothing -> False
+        , Require $ \input _ ->
+            isNothing (currentBid input)
+        ]
+
+enoughForFees :: Int -> Bool
+enoughForFees = (> 1_000_000)
+
+data End (v :: Type -> Type)
+  = End
+      (Var PubKeyHash v)
+      (Var PubKeyHash v)
+      Int
+      (Var TxOutRef v)
+  deriving stock (Eq, Show, Generic)
+
+instance FunctorB End
+instance TraversableB End
+
+end :: forall m. MonadGen m => Command m (PropertyT RunIO) AuctionState
+end =
+  let gen :: AuctionState Symbolic -> Maybe (m (End Symbolic))
+      gen s = case currentBid s of
+        Just (owner', won', amt, ref) -> do
+          owner <- snd <$> Map.lookup owner' (users s)
+          won <- snd <$> Map.lookup won' (users s)
+          pure $ pure $ End owner won amt ref
+        _ -> Nothing
+
+      execute :: End Concrete -> PropertyT RunIO ()
+      execute
+        ( End
+            (Var (Concrete owner))
+            (Var (Concrete winer))
+            amt
+            (Var (Concrete ref))
+          ) = liftRun $ PSM.end owner winer amt ref
+   in Command
+        gen
+        execute
+        [ Update $ \s _i _o ->
+            let
+              (o, w, amt, _) = fromMaybe (error "end with no current bid") (currentBid s)
+             in
+              s
+                { currentBid = Nothing
+                , winner = Just w
+                , users =
+                    users s
+                      -- take money from winner
+                      & Map.updateWithKey
+                        (\_ (bal, pkh) -> Just (bal - amt, pkh))
+                        w
+                      -- add money to owner
+                      & Map.updateWithKey
+                        (\_ (bal, pkh) -> Just (bal + amt, pkh))
+                        o
+                }
+        , Require $ \input (End ownerPkh1 _ amt1 _) -> isJust $ do
+            (ownerU, _, amt2, _) <- currentBid input
+            ownerPkh2 <- snd <$> Map.lookup ownerU (users input)
+            guard $
+              ownerPkh2 == ownerPkh1
+                && amt1 == amt2
+        ]
+
+type RunIO = StateT Mock IO
+
+newtype Check (v :: Type -> Type)
+  = Check [Var PubKeyHash v]
+  deriving stock (Eq, Show, Generic)
+
+instance FunctorB Check
+instance TraversableB Check
+
+-- TODO afaict this has to be a command
```

Comment:
In hedgehog, indeed you need a command to query system state see https://qfpl.io/posts/intro-to-state-machine-testing-2/#:~:text=GetPlayerCount%20is%20an%20input%20with%20no%20arguments%20%E2%80%94%20as%20the%20name%20suggests%20it%E2%80%99s%20just%20a%20getter.%20It%20also%20didn%E2%80%99t%20appear%20in%20our%20state%20machine%20diagram.%20The%20reason%20we%E2%80%99ve%20included%20it%20now%20is%20that%20we%20need%20some%20sort%20of%20side%20channel%20to%20interrogate%20the%20relevant%20part%20of%20the%20application%20state. - this is one of the reasons we're moving away from it